tools/cli-generator/generate.go

```diff
@@ -0,0 +1,281 @@
+// limitations under the License.
+
+package main
+
+import (
+	"embed"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	//go:embed spec.yaml
+	spec []byte
+	//go:embed templates/*.txt
+	templateFolder embed.FS
+
+	templateFuncs = template.FuncMap{
+		"Year": func() int {
+			return time.Now().Year()
+		},
+		"Now": func() string {
+			return time.Now().Format(time.RFC3339)
+		},
+		"Join": strings.Join,
+	}
+
+	ErrGenerateCli = errors.New("error generating cli")
+)
+
+type CLI struct {
+	Commands []Command
+	Stores   []Store
+
+	overwrite bool
+	basePath  string
+	templates *template.Template
+}
+
+type Command struct {
+	CommandPath    string    `yaml:"command_path,omitempty"`
+	PackageName    string    `yaml:"package_name,omitempty"`
+	OutputTemplate string    `yaml:"output_template,omitempty"`
+	Description    string    `yaml:"description,omitempty"`
+	StoreName      string    `yaml:"store_name,omitempty"`
+	StoreMethod    string    `yaml:"store_method,omitempty"`
+	Template       string    `yaml:"template,omitempty"`
+	IDDescription  string    `yaml:"id_description,omitempty"`
+	IDName         string    `yaml:"id_name,omitempty"`
+	Example        string    `yaml:"example,omitempty"`
+	RequestType    string    `yaml:"request_type,omitempty"`
+	SubCommands    []Command `yaml:"sub_commands,omitempty"`
+}
+
+func (c *Command) TemplateFile() string {
+	return c.Template + ".txt"
+}
+
+func (c *Command) TemplateUnitTestFile() string {
+	return c.Template + "_test.txt"
+}
+
+type Store struct {
+	BaseFileName string          `yaml:"base_file_name,omitempty"`
+	Template     string          `yaml:"template,omitempty"`
+	Lister       *StoreInterface `yaml:"lister,omitempty"`
+	Describer    *StoreInterface `yaml:"describer,omitempty"`
+	Creator      *StoreInterface `yaml:"creator,omitempty"`
+	Updater      *StoreInterface `yaml:"updater,omitempty"`
+	Deleter      *StoreInterface `yaml:"deleter,omitempty"`
+}
+
+func (s *Store) TemplateFile() string {
+	return s.Template + ".txt"
+}
+
+func (s *Store) InterfaceNames() string {
+	names := []string{}
+	if s.Lister != nil {
+		names = append(names, s.Lister.Name)
+	}
+	if s.Describer != nil {
+		names = append(names, s.Describer.Name)
+	}
+	if s.Creator != nil {
+		names = append(names, s.Creator.Name)
+	}
+	if s.Updater != nil {
+		names = append(names, s.Updater.Name)
+	}
+	if s.Deleter != nil {
+		names = append(names, s.Deleter.Name)
+	}
+	return strings.Join(names, ",")
+}
+
+type StoreInterface struct {
+	Name       string `yaml:"name,omitempty"`
+	Method     string `yaml:"method,omitempty"`
+	SDKMethod  string `yaml:"sdk_method,omitempty"`
+	ArgName    string `yaml:"arg_name,omitempty"`
+	ArgType    string `yaml:"arg_type,omitempty"`
+	ReturnType string `yaml:"return_type,omitempty"`
+}
+
+func (c *Command) CommandPaths() []string {
+	return strings.Split(c.CommandPath, " ")
```

Comment:
careful here, command is not only split but camelCase needs to be turn into snake_case

Reply:
I thought commands were always lowercase

Reply:
feel free to use https://pkg.go.dev/github.com/tangzero/inflector#Underscorize as we already have that lib

Reply:
> I thought commands were always lowercase

We default to snakeCase and generate aliases for lowercase and kebab-case

Reply:
also this is more about `baseFileName` where this method is used than the command path the user sees and types